File: src-tauri/src/apikeys.rs
// Las claves de API de los modelos.
//
// Una cuenta de Adeorq puede ser dos cosas muy distintas, y esto es la segunda:
//
//   · Un LOGIN: el CLI guarda su sesión en una carpeta, y esa carpeta es la
//     cuenta. Es lo que hay en accounts.rs, y es lo que gasta tu suscripción.
//   · Una CLAVE: no se instala ni se loguea, se guarda. Y factura por tokens.
//
// La diferencia importa porque se paga distinto. Con suscripción no puedes
// pasarte; con clave sí, y sin verlo, y por eso el chat lleva su contador.
//
// La clave se guarda cifrada en el Gestor de Credenciales de Windows y NO vuelve
// al front nunca: ni para enseñarla ni para arrancar un CLI con ella. Cuando una
// terminal necesita una, el front pide `@secreto:api:<proveedor>` y quien lo
// cambia por la clave de verdad es Rust, justo antes de crear el proceso (ver
// pty.rs). Así la clave no pasa por el WebView ni aparece en ningún estado.

use serde::Serialize;

use crate::secrets;

/// Bajo qué nombre vive la clave de un proveedor.
fn nombre(proveedor: &str) -> Result<String, String> {
    // Solo lo que puede ser un id de proveedor: esto compone el nombre de una
    // credencial del sistema, y no se construyen nombres con lo que llegue.
    if proveedor.is_empty()
        || proveedor.len() > 32
        || !proveedor.chars().all(|c| c.is_ascii_alphanumeric() || c == '-')
    {
        return Err("proveedor no válido".into());
    }
    Ok(format!("api:{proveedor}"))
}

#[derive(Serialize)]
pub struct EstadoClave {
    pub proveedor: String,
    /// Los cuatro últimos caracteres, que es lo único que se puede enseñar sin
    /// enseñar la clave: sirve para saber CUÁL pusiste, no para usarla.
    pub cola: String,
}

/// Guarda la clave de un proveedor. No se comprueba contra su API a propósito:
/// cada proveedor valida de una forma y algunos cobran por intentarlo. La que
/// sí se comprueba es la de OpenRouter, porque su endpoint de cuenta es gratis
/// (ver openrouter.rs).
#[tauri::command]
pub fn api_key_put(proveedor: String, clave: String) -> Result<(), String> {
    let clave = clave.trim();
    if clave.is_empty() {
        return Err("no has pegado ninguna clave".into());
    }
    secrets::put(&nombre(&proveedor)?, clave)
}

#[tauri::command]
pub fn api_key_forget(proveedor: String) -> Result<(), String> {
    secrets::forget(&nombre(&proveedor)?)
}

/// Qué claves hay guardadas. Devuelve la cola de cada una y nunca la clave.
#[tauri::command]
pub fn api_keys_estado(proveedores: Vec<String>) -> Vec<EstadoClave> {
    let mut out = Vec::new();
    for p in proveedores {
        let Ok(n) = nombre(&p) else { continue };
        if let Some(k) = secrets::get(&n) {
            let cola: String = k.chars().rev().take(4).collect::<Vec<_>>().into_iter().rev().collect();
            out.push(EstadoClave { proveedor: p, cola });
        }
    }
    out
}

File: src/lib/apikeys.ts
// Abrir un CLI con tu clave de API en vez de con tu login.
//
// Son dos formas de pagar lo mismo, y por eso se elige a mano y no se adivina:
// con tu suscripción no puedes pasarte, pagas igual hables mucho o poco; con
// clave se factura por tokens, y una tarde de pruebas es una factura.
//
// La clave NUNCA pasa por aquí. Ni al guardarla vuelve, ni al abrir la terminal
// se pide: lo que viaja es el MARCADOR `@secreto:api:<proveedor>`, y quien lo
// cambia por la clave de verdad es Rust, justo antes de crear el proceso (ver
// pty.rs y apikeys.rs). Así no aparece en el estado de React, ni en un log, ni
// en una captura de pantalla.

import { invoke } from "@tauri-apps/api/core";
import { providerOf } from "./providers";

export interface EstadoClave {
  proveedor: string;
  /** Los cuatro últimos caracteres: para saber CUÁL pusiste, no para usarla. */
  cola: string;
}

export function apiKeyPut(proveedor: string, clave: string): Promise<void> {
  return invoke("api_key_put", { proveedor, clave });
}

export function apiKeyForget(proveedor: string): Promise<void> {
  return invoke("api_key_forget", { proveedor });
}

export function apiKeysEstado(proveedores: string[]): Promise<EstadoClave[]> {
  return invoke("api_keys_estado", { proveedores });
}

/**
 * Dónde acaban las claves en ESTE sistema, para poder decirlo sin mentir.
 *
 * En Windows van al Gestor de Credenciales, cifradas con tu login. En Linux no
 * hay tal cosa: es un archivo con permisos `600`, que protege de otros usuarios
 * de la máquina pero no de otro programa tuyo. La pantalla prometía lo primero
 * en los dos sitios, y esa promesa hay que cumplirla o no hacerla.
 */
export type DondeSecretos = "credenciales" | "archivo";

export function secretosDonde(): Promise<DondeSecretos> {
  return invoke<DondeSecretos>("secretos_donde").catch(() => "archivo" as const);
}

/** Qué proveedores abren con clave en vez de con la suscripción. */
export const MODO_API_KEY = "adeorq-modo-api";

export const MODO_API_EVENTO = "adeorq:modo-api";

export function leerModoApi(): string[] {
  try {
    const raw = localStorage.getItem(MODO_API_KEY);
    const v = raw ? (JSON.parse(raw) as unknown) : [];
    return Array.isArray(v) ? v.filter((x): x is string => typeof x === "string") : [];
  } catch {
    return [];
  }
}

export function guardarModoApi(ids: string[]): void {
  localStorage.setItem(MODO_API_KEY, JSON.stringify(ids));
  window.dispatchEvent(new CustomEvent(MODO_API_EVENTO));
}

/**
 * El entorno con el que nace una terminal de este proveedor, o undefined si va
 * con tu suscripción, que es lo normal.
 *
 * Devuelve el marcador y no la clave: ver la cabecera de este archivo.
 */
export function entornoDe(proveedor: string): Record<string, string> | undefined {
  if (!leerModoApi().includes(proveedor)) return undefined;
  const v = providerOf(proveedor).apiEnv;
  if (!v) return undefined;
  return { [v]: `@secreto:api:${proveedor}` };
}
